app/com/arpnetworking/metrics/portal/query/impl/KairosDbQueryExecutor.java

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright 2020 Dropbox, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.arpnetworking.metrics.portal.query.impl;
+
+import com.arpnetworking.kairos.client.models.Aggregator;
+import com.arpnetworking.kairos.client.models.Metric;
+import com.arpnetworking.kairos.client.models.MetricsQueryResponse;
+import com.arpnetworking.kairos.client.models.Sampling;
+import com.arpnetworking.kairos.client.models.SamplingUnit;
+import com.arpnetworking.kairos.service.KairosDbService;
+import com.arpnetworking.metrics.portal.query.QueryExecutor;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Streams;
+import models.internal.BoundedMetricsQuery;
+import models.internal.MetricsQuery;
+import models.internal.MetricsQueryFormat;
+import models.internal.MetricsQueryResult;
+import models.internal.TimeSeriesResult;
+import models.internal.impl.DefaultMetricsQueryResult;
+import models.internal.impl.DefaultTimeSeriesResult;
+
+import java.time.temporal.ChronoUnit;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import javax.inject.Inject;
+
+/**
+ * A {@code QueryExecutor} that accepts KairosDB JSON metrics queries.
+ *
+ * @author Christian Briones (cbriones at dropbox dot com)
+ */
+public class KairosDbQueryExecutor implements QueryExecutor {
+    private final KairosDbService _service;
+    private final ObjectMapper _objectMapper;
+
+    /**
+     * Default Constructor.
+     *
+     * @param service The KairosDBService used to execute queries.
+     * @param objectMapper The objectMapper used to parse queries.
+     */
+    @Inject
+    public KairosDbQueryExecutor(final KairosDbService service, final ObjectMapper objectMapper) {
+        _service = service;
+        _objectMapper = objectMapper;
+    }
+
+    @Override
+    public CompletionStage<MetricsQueryResult> executeQuery(final BoundedMetricsQuery query) {
+        return CompletableFuture.completedFuture(query).thenCompose(this::executeQueryInner);
```

Comment:
I'd have thought that `CompletableFuture.completedFuture(x).thenCompose(y)` was semantically identical to `y(x)` -- what's the difference?

Reply:
There's a slight difference in error-handling semantics, although I'm unsure if there's a more explicit way to achieve the same effect.

The difference is that in the former, if y(x) throws an exception it is thrown "in the monad" so that the entire function returns a failed future containing said exception. If you instead called y(x) directly and it threw, that exception would be on the current thread of execution. I could get the same exception wrapping effect manually by putting a try/catch around y(x), performing the wrapping myself, but this current setup handles all of that for me behind the scenes.

i.e. the equivalent would be
```java
try {
    return executeQueryInner(query);
} catch (final Exception e) {
    final CompletableFuture<MetricsQueryResponse> future = new CompletableFuture<>();
    future.completeExceptionally(e);
    return future;
}
```

I noticed there's [CompletableFuture#supplyAsync](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#supplyAsync-java.util.function.Supplier-) but that runs it by default in the global fork-join pool. There's a separate executor argument you can specify but at that point I wouldn't know where to pull that from since ideally it would run in the same pool as the query itself.

Reply:
Future<T, E> 
Result<Future<T, E>, E>

Reply:
`supplyAsync` doesn't work because I want `executeQueryInner` to return a `CompletionStage`. I just rewrote it with the explicit try-catch.

Reply:
Ahhh, I understand now! (I suspect that, if you don't throw in a comment or a test for this, somebody will accidentally simplify this feature away. But it doesn't seem like a huge deal.)

Reply:
There's a test that catches this already!